Compiler regression for llvm-cpu target in IREE tag 3.8.0rc20250909

### What happened?

error: 'llvm.extractelement' op operand #0 must be LLVM dialect-compatible vector type, but got '!llvm.array<1 x vector<2xf32>>'
E                   %301 = torch.aten.scaled_dot_product_attention %298, %299, %300, %285, %float0.000000e00, %false_302, %none_303, %false_304 : !torch.vtensor<[32,8,1,32],f16>, !torch.vtensor<[32,8,?,32],f16>, !torch.vtensor<[32,8,?,32],f16>, !torch.vtensor<[32,1,1,?],f16>, !torch.float, !torch.bool, !torch.none, !torch.bool -> !torch.vtensor<[32,8,1,32],f16>


In SharkAI:
export IREE_SAVE_TEMPS=/home/dezhliao/shark-ai/IREE_ISSUE
sharktank/tests/models/llama/toy_llama_test.py::TestToyLlamaIree::testDecodePerplexity[False]


### Steps to reproduce your issue

[model.mlir.txt](https://github.com/user-attachments/files/22370640/model.mlir.txt)

iree-compile model.mlir\
             --iree-input-type=auto \
             --iree-vm-bytecode-module-output-format=flatbuffer-binary \
             --mlir-print-debuginfo \
             --mlir-print-op-on-diagnostic=false \
             --iree-hal-target-device=local \
             --iree-hal-local-target-device-backends=llvm-cpu


### What component(s) does this issue relate to?

_No response_

### Version information

_No response_

### Additional context

_No response_

--- a/compiler/src/iree/compiler/Codegen/LLVMCPU/ConvertToLLVM.cpp
+++ b/compiler/src/iree/compiler/Codegen/LLVMCPU/ConvertToLLVM.cpp
@@ -1015,6 +1015,7 @@ void ConvertToLLVMPass::runOnOperation() {
     vector::populateVectorMaskOpLoweringPatterns(patterns);
     vector::populateVectorShapeCastLoweringPatterns(patterns);
     vector::populateVectorFromElementsLoweringPatterns(patterns);
+    vector::populateVectorToElementsLoweringPatterns(patterns);
     // TODO: doubtful that the "default" does what one want here, it is likely
     // better to use shuffle.
     vector::populateVectorTransposeLoweringPatterns(
@@ -1091,6 +1092,7 @@ void ConvertToLLVMPass::runOnOperation() {
   populateVectorToLLVMConversionPatterns(typeConverter, patterns,
                                          reassociateFpReductions);
   vector::populateVectorFromElementsLoweringPatterns(patterns);
+  vector::populateVectorToElementsLoweringPatterns(patterns);
   ub::populateUBToLLVMConversionPatterns(typeConverter, patterns);
   vector::populateVectorTransferLoweringPatterns(patterns,
                                                  /*maxTransferRank=*/1);

--- a/tests/e2e/regression/BUILD.bazel
+++ b/tests/e2e/regression/BUILD.bazel
@@ -75,6 +75,7 @@ iree_check_single_backend_test_suite(
 iree_check_single_backend_test_suite(
     name = "check_regression_llvm-cpu",
     srcs = [
+        "dynamic_gather_attention.mlir",
         "layernorm.mlir",
         "lowering_config.mlir",
         "pack_pad_transpose_1x9_into_2x1x8x4_issue_12546.mlir",

--- a/tests/e2e/regression/BUILD.bazel
+++ b/tests/e2e/regression/BUILD.bazel
@@ -87,6 +87,20 @@ iree_check_single_backend_test_suite(
     target_backend = "llvm-cpu",
 )
 
+# TODO(22013): Merge the test suite into the main one, after the issue is fixed.
+iree_check_single_backend_test_suite(
+    name = "check_regression_dynamic_gather_attention_llvm-cpu",
+    srcs = [
+        "dynamic_gather_attention.mlir",
+    ],
+    compiler_flags = ["--iree-llvmcpu-target-cpu=generic"],
+    driver = "local-task",
+    tags = [
+        "noriscv",
+    ],
+    target_backend = "llvm-cpu",
+)
+
 iree_check_single_backend_test_suite(
     name = "check_regression_tosa_llvm-cpu",
     srcs = [

--- a/tests/e2e/regression/BUILD.bazel
+++ b/tests/e2e/regression/BUILD.bazel
@@ -142,6 +142,7 @@ iree_check_single_backend_test_suite(
 iree_check_single_backend_test_suite(
     name = "check_regression_hip",
     srcs = [
+        "dynamic_gather_attention.mlir",
         "linalg_ops_dynamic.mlir",
         "split_reduction_using_tiling.mlir",
     ],

--- a/tests/e2e/regression/BUILD.bazel
+++ b/tests/e2e/regression/BUILD.bazel
@@ -75,7 +75,6 @@ iree_check_single_backend_test_suite(
 iree_check_single_backend_test_suite(
     name = "check_regression_llvm-cpu",
     srcs = [
-        "dynamic_gather_attention.mlir",
         "layernorm.mlir",
         "lowering_config.mlir",
         "pack_pad_transpose_1x9_into_2x1x8x4_issue_12546.mlir",

--- a/tests/e2e/regression/CMakeLists.txt
+++ b/tests/e2e/regression/CMakeLists.txt
@@ -54,6 +54,7 @@ iree_check_single_backend_test_suite(
     "dynamic_abs.mlir"
     "dynamic_add.mlir"
     "dynamic_dot.mlir"
+    "dynamic_gather_attention.mlir"
     "dynamic_reduce_min.mlir"
     "dynamic_torch_index_select_high_rank.mlir"
     "dynamic_torch_index_select_negative.mlir"

--- a/tests/e2e/regression/CMakeLists.txt
+++ b/tests/e2e/regression/CMakeLists.txt
@@ -83,6 +83,21 @@ iree_check_single_backend_test_suite(
 
 iree_check_single_backend_test_suite(
   NAME
+    check_regression_dynamic_gather_attention_llvm-cpu
+  SRCS
+    "dynamic_gather_attention.mlir"
+  TARGET_BACKEND
+    "llvm-cpu"
+  DRIVER
+    "local-task"
+  COMPILER_FLAGS
+    "--iree-llvmcpu-target-cpu=generic"
+  LABELS
+    "noriscv"
+)
+
+iree_check_single_backend_test_suite(
+  NAME
     check_regression_tosa_llvm-cpu
   SRCS
     "linalg_quantized_matmul_vs_linalg_matmul.mlir"

--- a/tests/e2e/regression/CMakeLists.txt
+++ b/tests/e2e/regression/CMakeLists.txt
@@ -190,6 +190,7 @@ iree_check_single_backend_test_suite(
   NAME
     check_regression_hip
   SRCS
+    "dynamic_gather_attention.mlir"
     "linalg_ops_dynamic.mlir"
     "split_reduction_using_tiling.mlir"
   TARGET_BACKEND

--- a/tests/e2e/regression/CMakeLists.txt
+++ b/tests/e2e/regression/CMakeLists.txt
@@ -54,7 +54,6 @@ iree_check_single_backend_test_suite(
     "dynamic_abs.mlir"
     "dynamic_add.mlir"
     "dynamic_dot.mlir"
-    "dynamic_gather_attention.mlir"
     "dynamic_reduce_min.mlir"
     "dynamic_torch_index_select_high_rank.mlir"
     "dynamic_torch_index_select_negative.mlir"

--- a/tests/e2e/regression/dynamic_gather_attention.mlir
+++ b/tests/e2e/regression/dynamic_gather_attention.mlir
@@ -1,0 +1,26 @@
+#map = affine_map<(d0, d1, d2, d3, d4, d5, d6) -> (d0, d1, d2, d4)>
+#map1 = affine_map<(d0, d1, d2, d3, d4, d5, d6) -> (d0, d5, d1, d6, d4)>
+#map2 = affine_map<(d0, d1, d2, d3, d4, d5, d6) -> (d0, d5, d1, d6, d3)>
+#map3 = affine_map<(d0, d1, d2, d3, d4, d5, d6) -> ()>
+#map4 = affine_map<(d0, d1, d2, d3, d4, d5, d6) -> (d0, d1, d2, d5, d6)>
+#map5 = affine_map<(d0, d1, d2, d3, d4, d5, d6) -> (d0, d1, d2, d3)>
+func.func @gather_attention() {
+  %0 = util.unfoldable_constant dense<1.000000e+00> : tensor<32x4x2x32xf16>
+  %1 = flow.tensor.dynamic_constant dense<5.000000e-01> : tensor<2x4x16x32xf16> -> tensor<?x4x16x32xf16>
+  %2 = flow.tensor.dynamic_constant dense<1.500000e+00> : tensor<2x4x16x32xf16> -> tensor<?x4x16x32xf16>
+  %3 = flow.tensor.dynamic_constant dense<1> : tensor<32x2xi64> -> tensor<32x?xi64>
+  %4 = flow.tensor.dynamic_constant dense<1.500000e+00> : tensor<32x4x2x2x16xf16> -> tensor<32x4x2x?x16xf16>
+  %c1 = arith.constant 1 : index
+  %dim = tensor.dim %3, %c1 : tensor<32x?xi64>
+  %5 = tensor.empty(%dim) : tensor<32x?x4x16x32xf16>
+  %6 = iree_linalg_ext.gather dimension_map = [0] ins(%1, %3 : tensor<?x4x16x32xf16>, tensor<32x?xi64>) outs(%5 : tensor<32x?x4x16x32xf16>) -> tensor<32x?x4x16x32xf16>
+  %7 = iree_linalg_ext.gather dimension_map = [0] ins(%2, %3 : tensor<?x4x16x32xf16>, tensor<32x?xi64>) outs(%5 : tensor<32x?x4x16x32xf16>) -> tensor<32x?x4x16x32xf16>
+  %cst = arith.constant 1.767580e-01 : f16
+  %8 = tensor.empty() : tensor<32x4x2x32xf16>
+  %9 = iree_linalg_ext.attention {indexing_maps = [#map, #map1, #map2, #map3, #map4, #map5]} ins(%0, %6, %7, %cst, %4 : tensor<32x4x2x32xf16>, tensor<32x?x4x16x32xf16>, tensor<32x?x4x16x32xf16>, f16, tensor<32x4x2x?x16xf16>) outs(%8 : tensor<32x4x2x32xf16>) {
+  ^bb0(%arg0: f32):
+    iree_linalg_ext.yield %arg0 : f32
+  } -> tensor<32x4x2x32xf16>
+  check.expect_almost_eq_const(%9, dense<1.500000e+00> : tensor<32x4x2x32xf16>) : tensor<32x4x2x32xf16>
+  return
+}

